src/Interpreters/AsynchronousInsertQueue.cpp

```diff
@@ -374,6 +427,94 @@ AsynchronousInsertQueue::pushDataChunk(ASTPtr query, DataChunk chunk, ContextPtr
+    const auto max_ms = Milliseconds(settings.async_insert_busy_timeout_max_ms);
+
+    auto normalize = [&min_ms, &max_ms](const auto & t_ms) { return std::min(std::max(t_ms, min_ms), max_ms); };
+
+    if (!shard.last_insert_time || !flush_time_points.first)
+        return normalize(shard.busy_timeout_ms);
+
+    const auto & last_insert_time = *shard.last_insert_time;
+    const auto & [t1, t2] = std::tie(*flush_time_points.first, *flush_time_points.second);
+    const double increase_rate = settings.async_insert_busy_timeout_increase_rate;
+    const double decrease_rate = settings.async_insert_busy_timeout_decrease_rate;
+
+    const auto decreased_timeout_ms = std::min(
+        std::chrono::duration_cast<Milliseconds>(shard.busy_timeout_ms / (1.0 + decrease_rate)), shard.busy_timeout_ms - Milliseconds(1));
+
+    /// Increase the timeout for frequent inserts.
+    if (last_insert_time + min_ms > now)
+    {
+        auto timeout_ms = std::max(
+            std::chrono::duration_cast<Milliseconds>(shard.busy_timeout_ms * (1.0 + increase_rate)),
+            shard.busy_timeout_ms + Milliseconds(1));
+        if (timeout_ms != shard.busy_timeout_ms)
+            LOG_TRACE(
+                log,
+                "Async timeout increased from {} to {} for queue shard {}.",
+                shard.busy_timeout_ms.count(),
+                timeout_ms.count(),
+                shard_num);
+
+        return normalize(timeout_ms);
+    }
+    /// Decrease the timeout if inserts are not frequent,
+    /// that is, if the time since the last insert and the difference between the last two queue flushes were both
+    /// long enough (exceeding the adjusted timeout).
+    /// This ensures the timeout value converges to the minimum over time for non-frequent inserts.
+    else if (last_insert_time + decreased_timeout_ms < now && t1 + decreased_timeout_ms < t2)
+    {
+        auto timeout_ms = decreased_timeout_ms;
+        if (timeout_ms != shard.busy_timeout_ms)
+            LOG_TRACE(
+                log,
+                "Async timeout decreased from {} to {} for queue shard {}.",
+                shard.busy_timeout_ms.count(),
+                timeout_ms.count(),
+                shard_num);
+
+        return normalize(timeout_ms);
+    }
+
+    return normalize(shard.busy_timeout_ms);
+}
+
+void AsynchronousInsertQueue::validateSettings(const Settings & settings)
+{
+    const auto t_ms = std::chrono::milliseconds(settings.async_insert_busy_timeout_max_ms);
+
+    if (t_ms == std::chrono::milliseconds::zero())
+        throw Exception(ErrorCodes::INVALID_SETTING_VALUE, "Setting 'async_insert_busy_timeout_ms' can't be zero");
+
+    if (!settings.async_insert_use_adaptive_busy_timeout)
+        return;
+
+    /// Adaptive timeout settings.
+    const auto min_ms = std::chrono::milliseconds(settings.async_insert_busy_timeout_min_ms);
+
+    if (min_ms > t_ms)
```

Comment:
what will we do then for backward compatibility if someone set `async_insert_busy_timeout_ms` to a value less then default `async_insert_busy_timeout_min_ms`?

Reply:
Yes, it slipped my mind to update it, thanks. In that case, the minimum value of the two is taken as the async timeout.